File: apps/backend/app/contracts/board.py
"""Visual channel: BoardSnapshot, Element, VisionInterpretation.

(Architecture Document §6.3, §6.4). The board snapshot is captured by the
frontend, sent to the backend, then Vision turns it into a structured
interpretation that becomes the core input for the Learner.
"""

from __future__ import annotations

from pydantic import Field

from .base import CamelModel
from .enums import ElementType


class BoardSnapshot(CamelModel):
    """A board capture for one turn, sent from frontend to backend (§6.3)."""

    snapshot_id: str = Field(description="Unique snapshot identifier")
    session_id: str = Field(description="Session that owns the snapshot")
    turn_index: int = Field(description="Turn index (starts at 0)")
    image: str = Field(description="Image data as base64 or object URL")
    format: str = Field(description='Image format, e.g. "png"')
    captured_at: str = Field(description="Capture time (ISO-8601)")


class Element(CamelModel):
    """A single element detected on the board (§6.4).

    bbox is optional: [x, y, w, h].
    """

    type: ElementType
    content: str
    bbox: list[float] | None = Field(
        default=None, description="Bounding box [x, y, w, h]"
    )


class VisionInterpretation(CamelModel):
    """Vision's output over a snapshot — the core input for the Learner (§6.4).

    When confidence is below threshold, needsConfirmation is true and Vision
    includes a suggestedClarification instead of guessing silently (invariant §3.4).
    """

    snapshot_id: str = Field(description="Source snapshot")
    transcribed_text: str = Field(description="Text read from the board")
    elements: list[Element] = Field(
        default_factory=list, description="Detected elements"
    )
    confidence: float = Field(description="Confidence level 0..1")
    needs_confirmation: bool = Field(
        description="true when confidence is below threshold"
    )
    suggested_clarification: str | None = Field(
        default=None, description="Clarifying question for the user"
    )

File: apps/backend/src/contracts/board.ts
/**
 * Visual channel: BoardSnapshot, Element, VisionInterpretation
 * (Architecture Document §6.3, §6.4).
 *
 * The board snapshot is captured by the frontend, sent to the backend, then
 * Vision turns it into a structured interpretation that becomes the core input
 * for the Learner.
 */

import { z } from "zod";
import { elementTypeSchema } from "./enums.js";

/** A board capture for one turn, sent from frontend to backend (§6.3). */
export const boardSnapshotSchema = z.object({
  snapshotId: z.string(),
  sessionId: z.string(),
  turnIndex: z.number().int(),
  /** Image data as base64 or object URL. */
  image: z.string(),
  format: z.string(),
  capturedAt: z.string(),
});
export type BoardSnapshot = z.infer<typeof boardSnapshotSchema>;

/** A single element detected on the board (§6.4). bbox is optional: [x, y, w, h]. */
export const elementSchema = z.object({
  type: elementTypeSchema,
  content: z.string(),
  bbox: z.tuple([z.number(), z.number(), z.number(), z.number()]).optional(),
  /**
   * How sure Vision is about this one element, 0..1. Optional because not
   * every element comes from a model: the typed-text fallback builds elements
   * by hand and has no confidence to report.
   */
  confidence: z.number().optional(),
});
export type Element = z.infer<typeof elementSchema>;

/**
 * Vision's output over a snapshot — the core input for the Learner (§6.4).
 *
 * When confidence is below threshold, needsConfirmation is true and Vision
 * includes a suggestedClarification instead of guessing silently (invariant §3.4).
 */
export const visionInterpretationSchema = z.object({
  snapshotId: z.string(),
  transcribedText: z.string(),
  elements: z.array(elementSchema).default([]),
  /** Confidence level 0..1. */
  confidence: z.number(),
  /** true when confidence is below threshold. */
  needsConfirmation: z.boolean(),
  suggestedClarification: z.string().optional(),
  /**
   * What was added to the board this turn, read on its own.
   *
   * `transcribedText` is the whole board, old material included, because every
   * turn sends the whole board. On its own it cannot tell the student which part
   * the teacher just explained, and a student handed the whole board as "what
   * the teacher just taught" kept asking about topics the lesson had moved on
   * from. This is that part, read from an image of only the new strokes.
   *
   * Absent when there is nothing to compare against (the first reading of a
   * board, where everything is new); empty when the board did not change.
   */
  newText: z.string().optional(),
});
export type VisionInterpretation = z.infer<typeof visionInterpretationSchema>;
